firewood/src/merkle/stream.rs

```diff
@@ -485,19 +489,179 @@ mod tests {
     use test_case::test_case;
 
     #[tokio::test]
-    async fn iterate_empty() {
+    async fn key_value_iterate_empty() {
+        let merkle = create_test_merkle();
+        let root = merkle.init_root().unwrap();
+        let stream = merkle.key_value_iter_from(root, b"x".to_vec().into_boxed_slice());
+        check_stream_is_done(stream).await;
+    }
+
+    #[tokio::test]
+    async fn node_iterate_empty() {
         let merkle = create_test_merkle();
         let root = merkle.init_root().unwrap();
-        let stream = merkle.iter_from(root, b"x".to_vec().into_boxed_slice());
+        let stream = merkle.node_iter(root);
+        check_stream_is_done(stream).await;
+    }
+
+    #[tokio::test]
+    async fn node_iterate_root_only() {
+        let mut merkle = create_test_merkle();
+
+        let root = merkle.init_root().unwrap();
+
+        merkle.insert(vec![0x00], vec![0x00], root).unwrap();
+
+        let mut stream = merkle.node_iter(root);
+
+        let (key, node) = stream.next().await.unwrap().unwrap();
+
+        assert_eq!(key, vec![0x00].into_boxed_slice());
+        assert_eq!(node.inner().as_leaf().unwrap().data.to_vec(), vec![0x00]);
+
         check_stream_is_done(stream).await;
     }
 
+    #[tokio::test]
+    async fn node_iterator() {
+        let mut merkle = create_test_merkle();
+
+        let root = merkle.init_root().unwrap();
+
+        merkle
+            .insert(vec![0x00, 0x00, 0x00], vec![0x00, 0x00, 0x00], root)
+            .unwrap();
+        merkle
+            .insert(
+                vec![0x00, 0x00, 0x00, 0x01],
+                vec![0x00, 0x00, 0x00, 0x01],
+                root,
+            )
+            .unwrap();
+        merkle
+            .insert(
+                vec![0x00, 0x00, 0x00, 0xFF],
+                vec![0x00, 0x00, 0x00, 0xFF],
+                root,
+            )
+            .unwrap();
+        merkle
+            .insert(vec![0x00, 0xD0, 0xD0], vec![0x00, 0xD0, 0xD0], root)
+            .unwrap();
+        merkle
+            .insert(vec![0x00, 0xFF], vec![0x00, 0xFF], root)
+            .unwrap();
+
+        // Tree should be:
+        //     sentinel
+        //        | 0
+        //        00 <-- branch with no value
+        //     0/  D|   \F
+        //    00   0D0   F <-- leaf with no partial path
+        //  0/ \F
+        //  1   F
+        //
+        // Note the 0000 branch has no value and the F0F0
+        // The number next to each branch is the position of the child in the branch's children array.
+
+        // Test cases for when the iterator has no start key
```

Comment:
I strongly suggest you split these test cases up so that they are run as separate tests instead of checking different cases all in the same test. You can always have a separate function to create the same trie, but having a failing test case prevent a subsequent test does not lead to a great debugging experience.